---
title: "Shrubs Multivariate Analysis May 16"
author: "Jason Fridley"
date: "5/16/2020"
output: html_document
---
  
  ```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Dataset

This is a dataset of 68 quantitative traits for 76 species, plus 8 columns of categorical variables (nativity, invasive status [naturalized or invasive], stem ring porosity class, growth form, family, and phylogenetic group) and labels (taxon, species code). More information on traits is in Table S2. 

```{r echo=T}
traits = read.csv("/Users/fridley/Google Drive/shrubs-multivariate/shrub_traits_final.csv",row.names=1)
str(traits)
trcont = traits[,sapply(traits,class)!="factor"]  #continuous traits only
sum(is.na(trcont))/(dim(trcont)[1]*dim(trcont)[2]) #29% of values in the trait matrix are NAs
```

Inspection of distributions suggests the majority of traits show substantial right skew, which are log-transformed prior to analysis to reduce leverage of high trait values in linear models (eg PCA):

```{r echo=T, fig.show='hide'}
for(i in 1:dim(trcont)[2]) {  #figures not shown here
  par(mfrow=c(1,2),mar=c(5,5,1,1))
  plot(density(trcont[,i],na.rm=T),col="black",main=names(trcont[i]))
  plot(density(log(trcont[,i]),na.rm=T),col="red2",main=paste("log ",names(trcont[i]),sep=""))
}
```

``` {r echo=T}
#columns of high right skew to log-transform
log.cols = c(8,9,11,12,14,17,23,24,26,28,29,30,33,36,37,40,41:45,47:49,51,53:57,59,63,64,66,67,70,72,76)  
traits2 = traits  
for(i in log.cols) { traits2[,i] = log(traits2[,i]) }  #traits2 is trait matrix included logged columns
```


## Principal Components Analysis

The first objective is to examine major axes of trait variation across species, how traits relate to each other in terms of hypothesized trait spectra (eg, leaf economics), and whether native and invasive species occupy different regions of trait space. The linear form of this analysis for quantitative variables is usually PCA. However, because trait matrices often have more than a few missing values, there is a first step of filling in NAs ('imputation') using existing correlations (e.g., multivariate normal) in the dataset before principal components are calculated. There are several common ways to perform multivariate imputation in R, including reviews by Stacklies et al. (2007) and Dray and Josse (2014). Candidate methods that have been shown to perform well for datasets of the size and frequency of missing values as the current trait matrix include Bayesian PCA (BPCA) and probabilistic PCA (PPCA) in the 'pcaMethods' package (Stacklies et al. 2007), the joint modeling (JointM) method in the 'amelia' package (Honaker et al. 2011), and the iterative PCA (IPCA) method in the 'missMDA' package (Husson and Josse 2010). Of these, PPCA was most straightforward to use on the present dataset because it did not require choice of strong posteriors (as opposed to JointM) and was stable across the chosen number of PCs to use in imputation (as opposed to IPCA and BPCA). PPCA has also been shown to perform without bias in sparse datasets of up to 30% missing values (Grabich et al. 2014). Further, because imputed values of PPCA are fit closer to the center of the empirical distribution of each trait than the other methods (Stacklies et al. 2007), PPCA is a more conservative approach that in effect downweights the importance of traits of a high frequency of missing values.

PPCA is performed in R in the pcaMethods library, installed via the Bioconductor platform (<https://www.bioconductor.org/install/>). 

```{r message=F}
library(pcaMethods)    #https://academic.oup.com/bioinformatics/article/23/9/1164/272597
```

As input, we use a matrix of the continuous 68 traits for 76 species, including those log-transformed, and standardize columns to zero mean and unit variance.

```{r}
pcamat = scale(traits2[,sapply(traits,class)!="factor"])  
```

To determine how many principal components are to be used to impute missing values, pcaMethods includes the 'kEstimate' function that calculates error via cross validation, with choices of two different types of error (NRMSEP, normalized root mean square error, and Q2 distance, more suitable for smaller datasets; Stacklies et al. 2007). In this case both error methods give the same answer of two PCs. (This process takes about 15 min.)

```{r results=F}
#not run:
#estimate = kEstimate(pcamat,method="ppca",em="q2",evalPcs=1:10)
#estimate$bestNPcs  #optimal number of PCs = 2
```

We now run PPCA on out standardized trait matrix, returning the first two PCs. I include a random seed to set results 'in stone', but results for me were nearly identical regardless of seed.

```{r}
ppca1 = pcaMethods::pca(pcamat,method="ppca",seed=123,nPcs=2)
ppca1    #PPCA eigenvalues: PC1 = 18%, PC2 = 14% (32% of total variance)
#loadings(ppca1)[,1]  #access trait loadings for PC1
#ppca1.s1 = scores(ppca1)[,1]    #access species scores for PC1
#ppca1.s2 = scores(ppca1)[,2]    #access species scores for PC2
impmat = pcaMethods::completeObs(ppca1) #complete dataset (with imputed values for NAs)
```

Before plotting the PCA it is important to ensure the distributions of the complete trait dataset are consistent with the observed values. Recall PPCA is conservative, so we expect missing values to be more near the center of the empirical distribution (black curves will be tighter than blue curves for those variables with many NAs):

```{r fig.width=3, fig.height=2}
par(mar=c(2,2,1,1))
for(i in 1:dim(pcamat)[2]) {
  x = scale(na.exclude(pcamat[,i])) #observed values
  plot(density(impmat[,i]),main=colnames(pcamat)[i],xlab="",ylab="")
  lines(density(x),col="blue")
}
```

We also need to make pretty labels and consider color and symbol annotations. Expressions can be used for subscripts.

```{r}
pretty = names(traits)
pretty[8] = expression('Mass'[leaf])
pretty[9] = 'SLA'
pretty[11] = 'Lobe'
pretty[12] = expression('PA'[leaf])
pretty[13] = expression('Thick'[leaf])
pretty[14] = expression('Stomate'[size])
pretty[15] = expression('Stomate'[index])
pretty[16] = expression('Rd'[area])
pretty[17] = expression('Rd'[mass])
pretty[18] = expression('Asat'[area])
pretty[19] = expression('Asat'[mass])
pretty[20] = expression('Vc'[max])
pretty[21] = expression('J'[max])
pretty[24] = 'PNUE'
pretty[26] = expression('CC'[area])
pretty[27] = expression('CC'[mass])
pretty[28] = expression('N'[mass])
pretty[29] = expression('N'[area])
pretty[30] = expression('N'[litter])
pretty[31] = expression('N'[res])
pretty[32] = expression('C'[mass])
pretty[33] = expression('C'[area])
pretty[34] = expression('C'[litter])
pretty[41] = expression('Chl'[area])
pretty[42] = expression('Chl'[mass])
pretty[43] = 'Chl_a:b'
pretty[44] = expression('Decomp'[leaf])
pretty[45] = 'DNA'
pretty[46] = 'SRL'
pretty[49] = expression('N'[root])
pretty[50] = expression('C'[root])
pretty[51] = expression('Lignin'[root])
pretty[52] = expression('Decomp'[root])
pretty[53] = 'LAgain'
pretty[54] = expression('LeafRate'[max])
pretty[55] = expression('NPP'[root])
pretty[56] = expression('SS'[root])
pretty[57] = expression('SS'[stem])
pretty[58] = expression('Starch'[root])
pretty[59] = expression('Starch'[stem])
pretty[60] = expression('NSC'[root])
pretty[61] = expression('NSC'[stem])
pretty[63] = expression('Vessel'[freq])
pretty[64] = expression('Metaxylem'[diam])
pretty[65] = expression('Earlywood'[diam])
pretty[66] = expression('Latewood'[diam])
pretty[67] = expression('Cond'[stem])
pretty[68] = expression('LeafFall'[50])
pretty[69] = expression('Lifespan'[leaf])
pretty[70] = 'EmergeDuration'
pretty[71] = 'LeafDuration'
pretty[72] = 'LeafTurnover'
pretty[73] = 'FallDuration'
pretty[74] = expression('Chl_sun'[50])
pretty[75] = expression('Chl_shade'[50])
pretty[76] = 'LeafOut'
pretty2 = pretty[sapply(traits,class)!="factor"] #subset of continuous traits
```

Traits can be colored according to tissue type.

```{r}
pretty.col = rep("darkgreen",length=dim(traits)[2])
pretty.col[c(46:52,55,56,58,60)] = "black"  #root traits
pretty.col[c(57,59,61:67)] = "brown"        #stem traits
pretty.col[c(45,53:54,68,70:73,76)] = "darkorange2" #whole-plant traits
table(pretty.col)
pretty.col2 = pretty.col[sapply(traits,class)!="factor"] #subset of continuous traits
```

Species can be colored according to nativity status (blue = native, red = invasive, purple = naturalized), including a transparency value of 50% for the case of overlapping points. We also create transparency versions for trait colors.

```{r}
#col2rgb("red3") #205,0,0
#col2rgb("blue") #0,0,255
#col2rgb("purple") #160,32,240
color1 = rgb(205/255,0,0,.5) #"#EE000080"
color2 = rgb(0,0,1,.5) #"#0000FF80"
color3 = rgb(160/255,32/255,240/255,.5) #"#A020F080"

#col2rgb("darkgreen") #0,100,0
#col2rgb("brown") #165,42,42
#col2rgb("black") #0,0,0
#col2rgb("darkorange3") #205,102,0
color4 = rgb(0,100/255,0,.5) #"#00640080"
color5 = rgb(165/255,42/255,42/255,.5) #"#A52A2A80"
color6 = rgb(0,0,0,.5) # "#00000080"
color7 = rgb(205/255,102/255,0,.5) #"#CD660080"
trait.col = factor(pretty.col2)
levels(trait.col) = c(color6,color5,color4,color7)
trait.col= as.character(trait.col)
```

Here is a 2-panel figure of PPCA results, presented as Fig. 1. Note I've reversed PC1 just for visual purposes.

```{r fig.height=14, fig.width=7}
par(mfrow=c(2,1),mar=c(5,5,3,1))

#species scores: flip PC1
plot(-pcaMethods::scores(ppca1)[,1],pcaMethods::scores(ppca1)[,2],xlab="PC1 (18%)",ylab="PC2 (14%)",cex.lab=1.5,col=c(color1,color2,color3)[as.numeric(traits$status)],pch=19,cex=1.2)
abline(v=0,lty=2,col="gray"); abline(h=0,lty=2,col="gray")
points(-pcaMethods::scores(ppca1)[,1],pcaMethods::scores(ppca1)[,2],cex.lab=1.5,col=c(color1,color2,color3)[as.numeric(traits$status)],pch=21,cex=1.2,lwd=1.5)
abline(v=0,lty=2,col="gray"); abline(h=0,lty=2,col="gray")
text(-pcaMethods::scores(ppca1)[,1]+.2,pcaMethods::scores(ppca1)[,2]-.25,rownames(traits),col=c("red3","blue","blueviolet")[as.numeric(traits$status)],cex=.5,xpd=T)
mtext("Native (44)",side=3,at=-4,cex=1.2,line=.5,col="blue")
mtext("Naturalized (8)",side=3,at=0,cex=1.2,line=.5,col="blueviolet")
mtext("Invasive (24)",side=3,at=4,cex=1.2,line=.5,col="red3")

#trait scores
plot(-pcaMethods::loadings(ppca1)[,1],pcaMethods::loadings(ppca1)[,2],xlab="PC1 (18%)",ylab="PC2 (14%)",xlim=c(-.2,.26),ylim=c(-.27,.24),cex.lab=1.5,col=trait.col,pch=19,cex=1)
abline(v=0,lty=2,col="gray"); abline(h=0,lty=2,col="gray")
points(-pcaMethods::loadings(ppca1)[,1],pcaMethods::loadings(ppca1)[,2],xlim=c(-.6,.75),cex.lab=1,col=trait.col,pch=21,cex=1,lwd=1.5)
text(-pcaMethods::loadings(ppca1)[,1]+.01,pcaMethods::loadings(ppca1)[,2]-.01,pretty2,col=pretty.col2,cex=.65,xpd=T) 
mtext("Trait type:",side=3,at=-.14,cex=1.2,line=.5,col="gray21")
mtext("Leaf",side=3,at=-.06,cex=1.2,line=.5,col="darkgreen")
mtext("Stem",side=3,at=-0,cex=1.2,line=.5,col="brown")
mtext("Root",side=3,at=.06,cex=1.2,line=.5,col="black")
mtext("Whole-plant",side=3,at=.15,cex=1.2,line=.5,col="darkorange3")
```

It looks like invaders primarily occupy the lower right portion of the PCA, natives the upper left, and naturalized are intermediate. We can confirm with simple ANOVAs of nativity and invasive status for PCs 1 and 2.

```{r}
summary(lm(pcaMethods::scores(ppca1)[,1]~traits$nativity)) #Ho: PC1 scores do not vary by nativity
summary(aov(pcaMethods::scores(ppca1)[,1]~traits$status))  #Ho: PC1 scores do not vary by status (incl. naturalized)
TukeyHSD(aov(pcaMethods::scores(ppca1)[,1]~traits$status))    #Tukey HSD post hoc test
summary(lm(pcaMethods::scores(ppca1)[,2]~traits$nativity)) #Ho: PC2 scores do not vary by nativity
summary(aov(pcaMethods::scores(ppca1)[,2]~traits$status))  #Ho: PC2 scores do not vary by status (incl. naturalized)
TukeyHSD(aov(pcaMethods::scores(ppca1)[,2]~traits$status))    #Tukey HSD post hoc test
```

There are strong differences in the location of native and non-native species along both axes. If species are broken down instead by 3 groups, including non-invasive naturalized species, the sample size for inference is small (N=8) but tests suggest naturalized species are intermediate on PC1 and fit with invaders on PC2. In addition, natives appear to have a weak negative association of PC1 and PC2, but not the non-native groups:
  
``` {r}
summary(lm(pcaMethods::scores(ppca1)[,1]~pcaMethods::scores(ppca1)[,2],subset=traits$nativity=="native"))
summary(lm(pcaMethods::scores(ppca1)[,1]~pcaMethods::scores(ppca1)[,2],subset=traits$nativity=="alien"))
summary(lm(pcaMethods::scores(ppca1)[,1]~pcaMethods::scores(ppca1)[,2],subset=traits$status=="invasive"))
```

We can also ask whether natives and invaders occupy different regions of trait space overall, using a trait distance metric that is insensitive to NAs (Gower distance), and an ANOVA approach that accounts for nonindenpendence through permutation (PERMANOVA). Both methods are available in the *vegan* library. 

``` {r message=F}
library(vegan)
adon1 = adonis(pcamat~traits$nativity,method="gower",na.rm=T,permutations=9999)
adon1
adon2 = adonis(pcamat~traits$status,method="gower",na.rm=T,permutations=9999)
adon2
```


## Univariate trait contrasts in relation to nativity, controlling for phylogeny

Although the shrub collection was chosen with phylogenetic contrasts in mind and the naturalized and invasive species come from taxonomic groups that also contain many native species, it is nonetheless interesting to determine how much of the nativity difference of a particular trait could stem from invaders being taxonomically different overall than the native flora. This is a possibility because some of the native target species are very common in ENA but taxonomically unusual (Lindera, Dirca, Calycanthus, Hamamelis, Zanthoxylum, Clethra). We can thus compare the effect size of 'nativity' for a given trait with and without controlling for phylogenetic autocorrelation. The most common way to do this is via generalized least squares regression (or phylogenetic generalized least squares regression, PGLS) using the *nlme package* with a correlation structure that reflects phylogenetic relatedness. PGLS requires an aged tree that can return phylogenetic distance between species. We have an existing aged tree for all species based on the angiosperm megaphylogeny and the most recent phylogenetic analyses of specific groups:
  
  ![](finaltree.jpg)

Several packages are needed for importing the tree file, making sure it corresponds to our species list with correct labels, creating a correlation matrix from it, calculating phylogenetic signal, and finally running a PGLS:
  
```{r message=F, warning=F}
library(ape)
library(geiger)
library(nlme)
library(phytools)
library(treeplyr)
```

We read in the tree file (Newick text file, created previously), make sure names match taxon names in the traits dataset, and rearrange as needed.

```{r}
tree = read.tree(file="C:\\Users\\Fridley\\Google Drive\\shrubs-multivariate\\finaltree.tre")
traits3 = traits2 #includes log transformations
targets = sub(" ","_",traits3$taxon)
targets[27] = "Euonymus_hamiltonianus"
targets[58] = "Sambucus_nigra"
targets[39] = "Lonicera_japonica"
targets[49] = "Lonicera_tatarica"
targets[51] = "Lonicera_xbella"
targets[71] = "Viburnum_opulus"
targets[70] = "Viburnum_nudum"
targets[62] = "Symphoricarpos_albus"
rownames(traits3) = targets
name.check(tree,traits3)
PICmat = as.data.frame(make.treedata(tree,traits3)$dat) #trait matrix in order of tree tips
rownames(PICmat) = tree$tip.label
```

We'll now calculate the standardized trait difference for native and non-native (or native vs. invasive-only) species across all traits, either ignoring or including phylogenetic autocorrelation. We'll also calculate a measure of phylogenetic signal called Pagel's $\lambda$, which ranges from 0 (none) to 1 (maximum signal). We'll run these tests in a loop over each trait. Note this analysis ignores NAs (imputed values are not used).

```{r message=F, warning=F, results="hide"}
#output table will 2 cols per test x 4 tests (+phy signal) for 68 variables
unitable = matrix(0,nrow=dim(pcamat)[2],ncol=11)
colnames(unitable) = c("ols.nat.coef","ols.nat.se","ols.inv.coef","ols.inv.se","phy.nat.coef","phy.nat.se","phy.inv.coef","phy.inv.se","phy.sig","phy.sigP","Ninv")
unitable = as.data.frame(unitable)
rownames(unitable) = colnames(pcamat)
var.cols = c(1:76)[sapply(traits,class)!="factor"] #continuous trait columns in 'traits'

#loop over variables 1:68
for(i in 1:68) {
  y1 = scale(PICmat[,var.cols[i]])
  tree2 = drop.tip(tree,tip=c(1:length(y1))[is.na(y1)]) #ignores NAs
  nat2 = PICmat$nativity[!is.na(y1)]
  y = y1[!is.na(y1)]
  tree3 = drop.tip(tree,tip=c(1:length(y1))[is.na(y1)|PICmat$status=="naturalized"]) #ignores NAs
  inv2 = PICmat$nativity[!is.na(y1)&PICmat$status!="naturalized"]
  y2 = y1[!is.na(y1)&PICmat$status!="naturalized"]
  #OLS: linear coefficient of nativity for 2-group contrast
  unitable$ols.nat.coef[i] = summary(lm(y1~PICmat$nativity))$coef[2] #coef
  unitable$ols.nat.se[i] = summary(lm(y1~PICmat$nativity))[[4]][2,2] #se
  #same linear test but ignoring naturalized taxa
  unitable$ols.inv.coef[i] = summary(lm(y1~PICmat$nativity,subset=PICmat$status!="naturalized"))$coef[2] #coef
  unitable$ols.inv.se[i] = summary(lm(y1~PICmat$nativity,subset=PICmat$status!="naturalized"))[[4]][2,2] #se
  #phylogenetic signal from phytools
  unitable$phy.sig[i] = phylosig(tree=tree2, y, method="lambda", test=T)$lambda
  unitable$phy.sigP[i] = phylosig(tree=tree2, y, method="lambda", test=T)$P
  #PGLS from nlme (same as in phytools), P-ANOVA from phytools
  phy.nat = gls(y~nat2,correlation=corBrownian(phy=tree2),method="ML")
  unitable$phy.nat.coef[i] = summary(phy.nat)$coef[2]
  unitable$phy.nat.se[i] = summary(phy.nat)$tTable[2,2]
  #PGLS without naturalized taxa
  phy.inv = gls(y2~inv2,correlation=corBrownian(phy=tree3),method="ML")
  unitable$phy.inv.coef[i] = summary(phy.inv)$coef[2]
  unitable$phy.inv.se[i] = summary(phy.inv)$tTable[2,2]
  unitable$Ninv[i] = length(y2) #number of species when naturalized are excluded
}

```

Here is code for Fig. 2 that summarizes traits by how much they discriminate between native and non-native species, using both ordinary least squares (black symbols) and PGLS (gray), along with 95% CIs. Also included are the rankings of each trait on PCs 1 and 2, the number of species measured (N), and the estimated phylogenetic signal ($\lambda$). 

```{r fig.height=14, fig.width=8}
uni.ord = unitable[order(abs(unitable[,1])),] #sorted by OLS nativity coef value, biggest to least
N.per.trait = apply(pcamat,2,function(x)sum(!is.na(x))) #sample size per trait
sorted.PC1 = (69-rank(abs(pcaMethods::loadings(ppca1)[,1])))[order(abs(unitable[,1]))] #rank on PC1
sorted.PC2 = (69-rank(abs(pcaMethods::loadings(ppca1)[,2])))[order(abs(unitable[,1]))] #rank on PC2
lam.bold.vec = as.numeric(uni.ord$phy.sigP<0.05)+1 #bold all lambdas where P suggests >0
pc1.bold.vec = as.numeric(sorted.PC1<11)+1 #bold top 10 PC1 ranks
pc2.bold.vec = as.numeric(sorted.PC2<11)+1 #bold top 10 PC2 ranks

#create figure
par(oma=c(0,0,1,5.3),mar=c(0,0,0,0))
plot(-uni.ord[,1],1:68,axes=F,xlim=c(-1.5,2.5),ylim=c(1,67),pch=19,xlab="",ylab="") #flipped so that 'nonnativeness' is positive
segments(0,0,0,68,lty=2,col="gray")
arrows(-uni.ord[,1]-uni.ord[,2]*2,1:68,-uni.ord[,1]+uni.ord[,2]*2,1:68,angle=90,length=.01,code=3)
points(-uni.ord[,5],1:68-.4,xlim=c(-2,2),col="gray",pch=19)
arrows(-uni.ord[,5]-uni.ord[,6]*2,1:68-.4,-uni.ord[,5]+uni.ord[,6]*2,1:68-.4,angle=90,length=.01,code=3,col="gray")
points(-uni.ord[,1],1:68,xlim=c(-1.5,2.5),pch=19) 
mtext("Higher for Natives",side=3,at=-1,cex=1,col="blue",line=-.5)
mtext("No difference",side=3,at=0,cex=.8,col="darkgray",line=-1.2)
mtext("Higher for Non-natives",side=3,at=1.1,cex=1,col="red3",line=-.5)
mtext(pretty2[order(abs(unitable[,1]))],side=4,at=(1:68),line=-7,las=1,cex=.8,col=pretty.col2[order(abs(unitable[,1]))])
arrows(-.1,69.6,-.5,69.6,xpd=NA,lwd=3.5,col="blue",length=.15)
arrows(.1,69.6,.5,69.6,xpd=NA,lwd=3.5,col="red3",length=.15)
text(3.25,1:68,format(round(uni.ord$phy.sig,2),nsmall=2),xpd=NA,font=lam.bold.vec,cex=.75)
text(3.25,69.62,expression(lambda),cex=1.5,xpd=NA)
text(1.99,72.6,"Trait",cex=1.2,xpd=NA)
text(3,1:68,N.per.trait[order(abs(unitable[,1]))],xpd=NA,cex=.75)
text(3,69.5,expression(italic('N')),cex=1,xpd=NA)
text(2.5,69.5,"PC1",cex=1.1,xpd=NA)
text(2.75,69.5,"PC2",cex=1.1,xpd=NA)
text(2.5,1:68,sorted.PC1,xpd=NA,cex=.75,font=pc1.bold.vec)
text(2.75,1:68,sorted.PC2,xpd=NA,cex=.75,font=pc2.bold.vec)
```

And here is the same figure (Fig. S2) but comparing effect sizes of traits using only invasive non-native species:

```{r fig.height=14, fig.width=8}
uni.ord = unitable[order(abs(unitable[,3])),] #sorted by OLS nativity coef value (nat'd excluded), biggest to least
N.per.trait = unitable$Ninv #sample size per trait
sorted.PC1 = (69-rank(abs(pcaMethods::loadings(ppca1)[,1])))[order(abs(unitable[,1]))] #rank on PC1
sorted.PC2 = (69-rank(abs(pcaMethods::loadings(ppca1)[,2])))[order(abs(unitable[,1]))] #rank on PC2
lam.bold.vec = as.numeric(uni.ord$phy.sigP<0.05)+1 #bold all lambdas where P suggests >0
pc1.bold.vec = as.numeric(sorted.PC1<11)+1 #bold top 10 PC1 ranks
pc2.bold.vec = as.numeric(sorted.PC2<11)+1 #bold top 10 PC2 ranks

#create figure
par(oma=c(0,0,1,5.3),mar=c(0,0,0,0))
plot(-uni.ord[,3],1:68,axes=F,xlim=c(-1.5,2.5),ylim=c(1,67),pch=19,xlab="",ylab="") #flipped so that 'nonnativeness' is positive
segments(0,0,0,68,lty=2,col="gray")
arrows(-uni.ord[,3]-uni.ord[,4]*2,1:68,-uni.ord[,3]+uni.ord[,4]*2,1:68,angle=90,length=.01,code=3)
points(-uni.ord[,7],1:68-.4,xlim=c(-2,2),col="gray",pch=19)
arrows(-uni.ord[,7]-uni.ord[,8]*2,1:68-.4,-uni.ord[,7]+uni.ord[,8]*2,1:68-.4,angle=90,length=.01,code=3,col="gray")
points(-uni.ord[,3],1:68,xlim=c(-1.5,2.5),pch=19) 
mtext("Higher for Natives",side=3,at=-1,cex=1,col="blue",line=-.5)
mtext("No difference",side=3,at=0,cex=.8,col="darkgray",line=-1.2)
mtext("Higher for Invaders",side=3,at=1.1,cex=1,col="red3",line=-.5)
mtext(pretty2[order(abs(unitable[,3]))],side=4,at=(1:68),line=-7,las=1,cex=.8,col=pretty.col2[order(abs(unitable[,3]))])
arrows(-.1,69.6,-.5,69.6,xpd=NA,lwd=3.5,col="blue",length=.15)
arrows(.1,69.6,.5,69.6,xpd=NA,lwd=3.5,col="red3",length=.15)
text(3.25,1:68,format(round(uni.ord$phy.sig,2),nsmall=2),xpd=NA,font=lam.bold.vec,cex=.75)
text(3.25,69.62,expression(lambda),cex=1.5,xpd=NA)
text(1.99,72.6,"Trait",cex=1.2,xpd=NA)
text(3,1:68,N.per.trait[order(abs(unitable[,3]))],xpd=NA,cex=.75)
text(3,69.5,expression(italic('N')),cex=1,xpd=NA)
text(2.5,69.5,"PC1",cex=1.1,xpd=NA)
text(2.75,69.5,"PC2",cex=1.1,xpd=NA)
text(2.5,1:68,sorted.PC1,xpd=NA,cex=.75,font=pc1.bold.vec)
text(2.75,1:68,sorted.PC2,xpd=NA,cex=.75,font=pc2.bold.vec)
```

As expected, accounting for phylogenetic autocorrelation in the trait data does not have a large impact on the results, because non-natives are well represented across the phylogeny. We can test for this explicitly by estimating  $\lambda$ for nativity or status:
  
```{r}
phylosig(tree,as.numeric(PICmat$nativity),method="lambda",test=T)
phylosig(tree,as.numeric(PICmat$status),method="lambda",test=T)
```

We can also test for a phylogenetic signal of primary trait axes (PC1, PC2):

```{r}
pca12 = data.frame(-pcaMethods::scores(ppca1)[,1],pcaMethods::scores(ppca1)[,2])
rownames(pca12) = targets
name.check(tree,pca12)
PICmatpca = as.data.frame(make.treedata(tree,pca12)$dat) #trait matrix in order of tree tips
rownames(PICmatpca) = tree$tip.label
names(PICmatpca) = c("PC1","PC2")
phylosig(tree,as.numeric(PICmatpca$PC1),method="lambda",test=T)
phylosig(tree,as.numeric(PICmatpca$PC2),method="lambda",test=T)
```

Fig. S3 plots PC1 and PC2 scores in relation to phylogeny:

```{r fig.height=14, fig.width=10}
pca12b = data.frame(-pcaMethods::scores(ppca1)[,1],pcaMethods::scores(ppca1)[,2],traits$status)
rownames(pca12b) = targets
name.check(tree,pca12b)
PICmatpca2 = as.data.frame(make.treedata(tree,pca12b)$dat) #trait matrix in order of tree tips
names(PICmatpca2) = c("PC1","PC2","status")
stat.col = c("red3","blue","purple")[as.numeric(PICmatpca2$status)]
par(mfrow=c(1,3),mar=c(0,0,0,0),oma=c(3,1,1,1))
plot.phylo(tree,cex=1.3,tip.color=stat.col,no.margin=TRUE)
barplot(PICmatpca2$PC1,horiz=T,col=stat.col,cex.axis=1.5)
mtext("PC1",3,at=0,cex=1.7,line=-1.6)
barplot(PICmatpca2$PC2,horiz=T,xlim=c(-6,6),col=stat.col,cex.axis=1.5)
mtext("PC2",3,at=0,cex=1.7,line=-1.6)
```

And finally we can examine the overall correlation of trait and phylogenetic dissimilarity, using a Mantel test and the above Gower's distance:

```{r}
PICmat2 = PICmat[,sapply(PICmat,class)!="factor"]
vdphy = vegdist(scale(PICmat2),method="gower",na.rm=T,diag=T,upper=F)
phydist = cophenetic(tree)
pd = phydist[lower.tri(phydist)]
mantel(phydist,vdphy)
```


## Which traits predict invasiveness?

So far the associations we've examined between traits and invasiveness have been strictly linear, and have ignored trait interactions. Typical models of 'species invasiveness' use a more predictive approach that stem from decision tree analysis, where we don't care as much about the mechanistic relationship between traits and invasivenesss, only whether we can use traits to make a good predictive model of which species are invasive and which aren't. We start with a simple example of a classification tree from the *rpart* library, based on LeafDuration, the variable most associated with invasiveness in the above linear analysis. 

```{r message=F, warning=F, fig.width=6}
library(rpart)
tree1 = rpart(formula=nativity~leafduration,data=traits,method="class")
tree1
summary(tree1)
plot(tree1); text(tree1,use.n=T,xpd=NA)
```

The simplest classification tree we can make separates 76 species into two groups based on whether they maintain live leaves more than 191 days per year. Group 1 has 26 non-native and 8 native species, and group 2 has 36 natives and 6 non-native species. So this single partition based on one variable correctly separates species based on nativity at a rate of (26+36)/76 = 81.5% accuracy. (If we exclude the 8 naturalized species and only separate native and invasive species, the accuracy is 58/68=85%). 

Where do we go from here, given we have 67 more traits to use? We could throw all traits in the above rpart model, but a more robust approach is to use a machine learning procedure wherein thousands of models are created and summarized to create a 'boosted' tree that can substantially increase model accuracy. So-scalled boosted regression trees (BRT) have become a common approach for invasiveness models when there are many potentially interacting predictors. Here I follow the general guidelines of MacGregor et al. (2012, Div & Dist 18:1013) using the *gbm* package, using all traits at our disposal for classifying native vs. non-native species. Decision trees are robust to missing data so we can use the non-imputed dataset, and we don't have to worry about monotonic transformations of the predictors because they have no effect on the results of a tree model. However, below I use the log-transformed version because it is easier to examine results graphically.

The "boosting" component of BRT models involves stochasticity and requires a few set parameters. The *learning rate* (aka 'shrinkage') determines the importance of each consecutively produced tree to the final model. The *complexity* parameter controls the allowed order of interactions (e.g., complexity of 2 means only 2-way interations are considered). The *bag faction* is the proportion of observations that are used to fit a tree in each iteration, drawn at random without replacement. This data subsetting procedure protects against overfitting. Although having three parameters to set a priori is an annoyance, Elith et al. (2008 J Anim Ecol 77:802-813) suggest: 1) bag fraction between 50-75%, 2) relatively 'slow' learning rate, around 0.005; 3) convergence at a number of trees above 1000, which is an outcome of all three parameters. Below, I use these recommendations, while also opting for as simple a model structure as possible using only 2-way interactions. Of course, any final result needs to be robust to various parameter settings, and this is achieved only through simulation. A helpful vignette is [here.](https://rdrr.io/cran/dismo/f/inst/doc/brt.pdf) 

```{r fig.width=4, message=F, warning=F}
library(dismo)
library(gbm)
#make new dataframe that excludes naturalized species, includes logs, no categorical variables
traits4 = traits2[,sapply(traits,class)!="factor"]
invasive = as.numeric(as.numeric(traits$nativity)==1) #make 0-1 dependent variable
input = data.frame(invasive,traits4)  #make col 1 dependent variable
input = input[traits$status!="naturalized",] #remove naturalized-only species
set.seed(456)
gbm1 <- gbm.step(data=input, gbm.x = c(2:69), gbm.y = 1, family = "bernoulli", tree.complexity = 2, learning.rate = 0.005, bag.fraction = 0.7)
```

The above figure shows the optimal number of trees produced (green line) as determined by a cross validation procedure that records the lowest predicted deviance (red line). Rising deviance after the minimum indicates overfitting. To identify which predictors have the greatest weight in this final composite tree model, a *relative importance* statistic is calculated that measures the number of times a predictor is selected for splitting across all created trees, weighted by how much the model is improved with each split. Relative importances are scaled across all predictors so they sum to 100.

```{r}
summary(gbm1)
```

Here the top variables are, as expected, LeafDuration and its close associate LeafFall50, or the day at which half of leaves have senesced. There are a handful of other variables that are important, but the decline is steep and most variables appear to have little to no significance in predicting which species are invasive. The *dismo* package includes a cross-validation algorithm wereby the least important predictors are removed progressively and compared against the original fitted (complex) model; the details are in Appendix S2 of Elith et al. (2008).

```{r}
simple = gbm.simplify(gbm1)
```

This procedure has removed all predictors except for two: LeafDuration and DNA content. Let's rerun the BRT model with only these predictors (saved in the simplify object as preds.66):

```{r warning=F}
set.seed(789)
gbm2 <- gbm.step(data=input, gbm.x = simple$pred.list$preds.66, gbm.y = 1, family = "bernoulli", tree.complexity = 2, learning.rate = 0.005, bag.fraction = 0.7)
```

How good is the model with only two predictors? One way is to calculate the 'area under the curve' (AUC = receiver operating characteristic, or RoC), which gbm.step saves automatically for each cross validation (N=10) in the vector gbm2$cv.roc.matrix. A better way is to visualize the fitted values against the observed. Here are both:

```{r fig.width=6, fig.height=9}
mean(gbm2$cv.roc.matrix)
logit.fitted = log(gbm2$fitted/(1-gbm2$fitted))
plot(input[,1],logit.fitted,type="n",xlim=c(-.5,1.5),xlab="",xaxt="n")
mtext(c("Native","Invasive"),at=c(0,1),side=1,line=1,cex=1.5)
text(jitter(input[,1]),logit.fitted,rownames(input),cex=.7); abline(h=0,lty=2,col="gray")
```

The mean AUC from the cross-validation tests is 0.96, which is exceptionally high signal detection compared to noise. The figure shows why: of 68 species, only 5 were misclassified (LOSE for natives; CEOR, VIDI, EUEU, EUAL for invaders), for a total predictive accuracy of 93%. Another visualization tool is to examine the probability of species invasiveness according to our two predictors across trees (ms Fig. 3):

```{r fig.width=7, fig.height=10, message=F}
Nvec = 100
x.range = c(min(input$dna,na.rm=T),max(input$dna,na.rm=T))
y.range = c(min(input$leafduration,na.rm=T),max(input$leafduration,na.rm=T))
x.vec = seq(from=x.range[1],to=x.range[2],length=Nvec)
y.vec = seq(from=y.range[1],to=y.range[2],length=Nvec)
zpred = function(x,y) {predict.gbm(gbm2,newdata=list(leafduration=y,dna=x),n.trees=gbm2$n.trees,type="response")}
z.vec = outer(x.vec,y.vec,zpred)
persp(x.vec,y.vec,z.vec,theta=45,phi=35,shade=.4,xlab="",ylab="",zlab="",r=3,d=1,scale=T,border=NA,
	col="white",axes=T,ticktype="detailed",cex.axis=1.2,zlim=c(0,1))
text(-.375,-.0028,"P(Invasive)",cex=1.6,xpd=NA,srt=280)
text(.195,-.31,"Leaf duration (d)",cex=1.6,xpd=NA,srt=42.5)
text(-.181,-.307,"DNA content (log pg)",cex=1.6,xpd=NA,srt=-43)
```

For completeness, we can compare the above model of invasiveness to one that discriminates only by whether a species is native or non-native (N=76 species). The results are similar:

```{r warning=F}
traits5 = traits2[,sapply(traits,class)!="factor"]
invasive = as.numeric(as.numeric(traits$nativity)==1) #make 0-1 dependent variable
input2 = data.frame(invasive,traits5)  #make col 1 dependent variable
set.seed(456)
gbm3 <- gbm.step(data=input2, gbm.x = c(2:69), gbm.y = 1, family = "bernoulli", tree.complexity = 2, learning.rate = 0.005, bag.fraction = 0.7)
#summary(gbm3)
simple2 = gbm.simplify(gbm3)
  #same two variables kept, leafduration and dna
gbm4 <- gbm.step(data=input2, gbm.x = simple2$pred.list$preds.66, gbm.y = 1, family = "bernoulli", tree.complexity = 2, learning.rate = 0.005, bag.fraction = 0.7)
logit.fitted = log(gbm4$fitted/(1-gbm4$fitted))
plot(input2[,1],logit.fitted,type="n",xlim=c(-.5,1.5),xlab="",xaxt="n")
mtext(c("Native","Non-native"),at=c(0,1),side=1,line=1,cex=1.5)
text(jitter(input2[,1]),logit.fitted,rownames(input2),cex=.7); abline(h=0,lty=2,col="gray")
```

Out of 76 species, only 4 were incorrectly assigned using LeafDuration and DNA predictors (LOSE for natives; CEOR, EUEU, EUHA for non-natives), for 95% accuracy. Ignoring the complexity of BRT and going back to a simple classification tree approach from *rpart*, we find that BRT improves the accuracy by 6 species:

```{r}
tree2 = rpart(formula=nativity~leafduration+dna,data=traits,method="class")
tree2
summary(tree2)
plot(tree2); text(tree2,use.n=T,xpd=NA) #10 out of 76 species misassigned
```

## End

